Add option to save real tensors in TORCH_COMPILE_DEBUG repro

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138110

This pr adds a utility to try to try to construct the corresponding real tensor values of fake tensors by seeing if their meta storage is contained in the meta converter. 

Then, we are able to save real tensor values for fx_graph_runnable if `TORCH_COMPILE_DEBUG_SAVE_REAL=1` is set.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

Differential Revision: [D64502744](https://our.internmc.facebook.com/intern/diff/D64502744)

eellison committed:
[WIP] save real inputs in inductor repro

[ghstack-poisoned]

eellison committed:
Update on "[WIP] save real inputs in inductor repro"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

eellison committed:
Update on "[WIP] save real inputs in inductor repro"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64502744).

eellison committed:
Update on "[WIP] save real inputs in inductor repro"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D64502744](https://our.internmc.facebook.com/intern/diff/D64502744)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64502744).

eellison committed:
Update on "[WIP] save real inputs in inductor repro"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D64502744](https://our.internmc.facebook.com/intern/diff/D64502744)

[ghstack-poisoned]

eellison committed:
Update on "[WIP] save real inputs in inductor repro"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D64502744](https://our.internmc.facebook.com/intern/diff/D64502744)

[ghstack-poisoned]

eellison committed:
Update on "[WIP] save real inputs in inductor repro"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D64502744](https://our.internmc.facebook.com/intern/diff/D64502744)

[ghstack-poisoned]

ezyang commented on torch/_dynamo/debug_utils.py:
```diff
@@ -511,7 +511,7 @@ def symint(self, *args, **kwargs):
 # transferring around
 class InputReader:
-    def __init__(self, save_dir=None, *, pbar=None):
+    def __init__(self, save_dir=None, *, pbar=None, stable_hash=False):
```
Argument does nothing?

ezyang commented on torch/_inductor/debug.py:
```diff
@@ -473,7 +473,26 @@ def fx_graph(
+                    inputs,
+                    "inductor",
+                    save_dir=save_dir,
+                    stable_hash=stable_hash,
```
Not sure I understand this? You only ever pass True here when this kwarg is actually respected?

ezyang commented on torch/_subclasses/fake_utils.py:
```diff
@@ -75,6 +76,74 @@ def is_sdpa_error(func, idx, e):
+    ten_list: List[Union[FakeTensor, Any]]
+) -> List[Union[FakeTensor, torch.Tensor, Any]]:
+    """
+    Attempt to convert fake tensors to a corresponding real tensor with the correct underlying storage by looking up
```
Can you explain why you have to do something this complicated? A trivial alternate implementation would be to have fake tensor maintain weakref to the real tensor they were converted from and read it off from that. Why do we have to do something so fancy?

eellison replied:
Aot autograd will do weird things with views and such. this is more reliable.

eellison committed:
Update on "Add option to save real tensors in TORCH_COMPILE_DEBUG repro"


This pr adds a utility to try to try to construct the corresponding real tensor values of fake tensors by seeing if their meta storage is contained in the meta converter. 

Then, we are able to save real tensor values for fx_graph_runnable if `TORCH_COMPILE_DEBUG_SAVE_REAL=1` is set.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov rec

Differential Revision: [D64502744](https://our.internmc.facebook.com/intern/diff/D64502744)

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge
